### PR TITLE
Upgrade react native to v0.71.0

### LINKED ISSUE #5949: Investigate autolinking for packages that should be auto linked
Our metro config is pretty vanilla, and that is a good thing. But we have hit a corner case where we thought we were auto linked yet we had to manually link to see the results. 

I believe this is because of how we are resolving our dependencies at the moment. 

We should be able to adopt a metro config that has some of these elements (not necessarily these exactly) to achieve this. In particular, the `extraNodeModules` key within the resolver should be something we look at.

This way we can go back to unlinking the relinked dependencies.

<details>
<summary>Example config</summary>

```js
"workspaces": {
  "packages" : [...],
  "noHoist": ["**"] //keep node_modules in the children folders

const path = require("path");

module.exports = {
  transformer: {
    getTransformOptions: async () => ({
      transform: {
        experimentalImportSupport: false,
        inlineRequires: false,
      },
    }),
  },

  /**
  * Ensure any imports inside the shared folder resolve to the local
  * node_modules folder
  */
  resolver: {
    extraNodeModules: new Proxy({}, {
      get: (target, name) => path.join(process.cwd(), `node_modules/${name}`),
    }),
  },

   /**
   * Add our workspace roots so that react native can find the source code for
   * the included packages in the monorepo
   */
  projectRoot: path.resolve(__dirname),
  watchFolders: [
    path.resolve(__dirname, "../common"),
    path.resolve(__dirname, "../../node_modules")
  ]
};
```

</details>

### LINKED ISSUE #5959: Remove "ignore" list from LogBox (source/root.tsx)
- [ ] Failed prop type: Invalid prop `containerTagName` of type `object` supplied to `ReactMarkdown`, expected `function`
- [ ] Failed prop type: Invalid prop `containerTagName` of type `string` supplied to `ReactMarkdown`, expected `function`.
- [ ] MaskedViewIOS has been extracted from react-native core and will be removed in a future release. It can now be installed and imported from `@react-native-community/masked-view` instead of `react-native`. See https://github.com/react-native-community/react-native-masked-view
- [ ] `componentWillReceiveProps` has been renamed, and is not recommended for use. See https://fb.me/react-async-component-lifecycle-hooks for details.
- [ ] `componentWillMount` has been renamed, and is not recommended for use. See https://fb.me/react-async-component-lifecycle-hooks for details.

### PR DESCRIPTION
**Pull request by drewvolz**
## Upgrade from RN 0.68.5 ~> 0.71.0

- [Upgrade helper diff for 68.5 ~> 71.0 ](https://react-native-community.github.io/upgrade-helper/?from=0.68.5&to=0.71.0)
- [Upstream release notes](https://github.com/facebook/react-native/blob/main/CHANGELOG.md#v071)
- [Release blog announcement](https://reactnative.dev/blog/2023/01/12/version-071)
- Closes #5949
- Closes #5959

### Testing this out:
1. `git checkout drew/rn-71`
2. `npm i`
3. `cd ios && pod install` (with a handful of `repo-update` follow-ups)
4. `npm run start --reset-cache`
5. iOS: open .xcworkspace and build (you may have to comment out `npm run bundle-data` in the "bundle data files"  build phase, maybe my npm is not setup correctly)
6. Android: sync gradle, then build
7. Connect to metro

### 👀 Noteworthy
- Logbox ignore has been removed (we will now see our previously hidden logs)
- I removed the storyboard loading screen logic for iOS in the app delegate (but not the storyboard)

### 🎈 What looks to be working
- [Flipper](https://fbflipper.com)
- Navigation
- General network requests
- Components rendering / interacting

### ⚠️ What looks broken
- Search?
  - [x] Directory seems to work (tested on Android)
  - [x] Filter search (courses, a-to-z) seem broken (tested on Android)
- [x] Popover v5.0 fixes #6728 

### 📝 Todo
- [x] Merge #6572 into this branch
- [x] Remove `@types/react-native`
- [x] Lots of other dependencies should be upgraded
- [ ] Check that `ReactDatabaseSupplier` (async storage) on android is still working after changes to `MainActivity`
- [ ] How do we add both debug and release `ReactNativeFlipper.java` class? (to prevent flipper from running on prod)

### CODE REVIEW

**Review by hawkrives — APPROVED** [review]

**Comment by hawkrives** [comment]
Android: 

>   - In plugin 'com.facebook.react' type 'com.facebook.react.tasks.BundleHermesCTask' property 'entryFile' specifies file '/home/runner/work/AAO-React-Native/AAO-React-Native/android/app/index.js' which doesn't exist.

**Comment by drewvolz** [comment]
Build for Android 
Screenshot 2023-01-17 at 12 35 35 PM

Build for iOS (detox)
Screenshot 2023-01-17 at 12 36 24 PM

**Review by hawkrives — APPROVED** [review]

File: .gitignore [hunk 1]
```diff
@@ -22,6 +22,7 @@ DerivedData
 *.hmap
 *.ipa
 *.xcuserstate
+ios/.xcode.env.local
```
Comment (rye) [inline .gitignore]: Can we anchor both this and the following `ios/` entries to the root `/ios` directory?
```suggestion
/ios/.xcode.env.local
```

**Review by rye — COMMENTED** [review]
One gitignore suggestion and then a question: do we need to commit the `gradle-wrapper.jar` file?

File: .gitignore (hunk 1 above)
Comment (drewvolz) [reply .gitignore]: Sure!

Comment (drewvolz) [reply .gitignore]: Regarding `gradle-wrapper.jar`: was in the diff helper to upgrade, changes are unknown.

**Comment by hawkrives** [comment]
@rye I forget why, but committing the binary file is recommended by Gradle and RN… although it seems strange

File: package.json [hunk 2]
```diff
@@ -50,11 +50,12 @@
...
     "@hawkrives/react-native-alternate-icons": "0.6.2",
     "@react-native-async-storage/async-storage": "1.17.11",
-    "@react-native-community/clipboard": "1.5.1",
+    "@react-native-clipboard/clipboard": "1.11.1",
     "@react-native-community/datetimepicker": "6.7.2",
     "@react-native-community/netinfo": "9.3.7",
+    "@react-native-community/segmented-control": "2.2.2",
     "@react-native-picker/picker": "2.4.8",
-    "@react-navigation/bottom-tabs": "6.5.3",
+    "@react-navigation/bottom-tabs": "6.4.0",
```
Comment (hawkrives) [inline package.json]: Be sure not to keep this downgraded!

Comment (drewvolz) [reply package.json]: Oop! Good catch

File: android/app/build.gradle [hunk 3]
```diff
@@ -1,82 +1,58 @@
...
+
+    /* Hermes Commands */
+    //   The hermes compiler command to run. By default it is 'hermesc'
+    // hermesCommand = "$rootDir/my-custom-hermesc/bin/hermesc"
+    //
+    //   The list of flags to pass to the Hermes compiler. By default is "-O", "-output-source-map"
+    // hermesFlags = ["-O", "-output-source-map"]
+}
 
 def skipBundling = System.getenv().containsKey("SKIP_BUNDLING")
```
Comment (hawkrives) [inline android/app/build.gradle:57]: Was project.ext.react supposed to merge into the react {} block?

File: android/app/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -1,7 +1,6 @@
 <manifest
     xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:tools="http://schemas.android.com/tools"
-    package="com.allaboutolaf"
```
Comment (hawkrives) [inline android/app/src/main/AndroidManifest.xml:4]: No more [package]? Neat!

File: android/build.gradle [hunk 5]
```diff
@@ -1,10 +1,8 @@
-import org.apache.tools.ant.taskdefs.condition.Os
-
 // Top-level build file where you can add configuration options common to all sub-projects/modules.
 
 buildscript {
     ext {
-        buildToolsVersion = "31.0.0"
+        buildToolsVersion = "33.0.0"
```
Comment (hawkrives) [inline android/build.gradle:5]: Yay!

File: android/app/src/main/AndroidManifest.xml (hunk 4 above)
Comment (hawkrives) [reply android/app/src/main/AndroidManifest.xml:4]: Ah,  I see: it's deprecated!

```
package="io.sentry.react" found in source AndroidManifest.xml: /Users/rives/Projects/allaboutolaf/node_modules/@sentry/react-native/android/src/main/AndroidManifest.xml.
Setting the namespace via a source AndroidManifest.xml's package attribute is deprecated.
Please instead set the namespace (or testNamespace) in the module's build.gradle file, as described here: https://developer.android.com/studio/build/configure-app-module#set-namespace
This migration can be done automatically using the AGP Upgrade Assistant, please refer to https://developer.android.com/studio/build/agp-upgrade-assistant for more information.
```

**Comment by hawkrives** [comment]
ReactNativeFlipper.initializeFlipper(this, getReactNativeHost().getReactInstanceManager());

… can't be found in release mode. Hrm. Is this what you were talking about with ensuring that flipper isn't included in release, @drewvolz ?

**Comment by drewvolz** [comment]
The diff checker shows two files -- one for release, one for debug. Finding out how to replicate this and add it

1. Debug: android/app/src/debug/java/com/rndiffapp/ReactNativeFlipper.java

2. Release: android/app/src/release/java/com/rndiffapp/ReactNativeFlipper.Java

```java
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * <p>This source code is licensed under the MIT license found in the LICENSE file in the root
 * directory of this source tree.
 */
package com.allaboutolaf;

import android.content.Context;
import com.facebook.react.ReactInstanceManager;

/**
 * Class responsible of loading Flipper inside your React Native application. This is the release
 * flavor of it so it's empty as we don't want to load Flipper.
 */
public class ReactNativeFlipper {
  public static void initializeFlipper(Context context, ReactInstanceManager reactInstanceManager) {
    // Do nothing as we don't want to initialize Flipper on Release.
  }
}
```

I know we have
- android/app/src/debug and 
- android/app/src/main

These are our equivalents, so if we can match it somehow... I think I was running into an error where the flipper classes were clashing when I tried to add a java/com/allaboutolaf/ directory under debug with a flipper file.

**Pull request merged**